cmd/soroban-rpc/internal/methods/get_ledgers.go

```diff
@@ -0,0 +1,241 @@
+	switch {
+	case req.Pagination != nil:
+		switch {
+		case req.Pagination.Cursor != "" && req.StartLedger != 0:
+			return errors.New("startLedger and cursor cannot both be set")
+		case req.Pagination.Limit > maxLimit:
+			return fmt.Errorf("limit must not exceed %d", maxLimit)
+		}
+	case req.StartLedger != 0 && !isStartLedgerWithinBounds(req.StartLedger, ledgerRange):
+		return fmt.Errorf(
+			"start ledger must be between the oldest ledger: %d and the latest ledger: %d for this rpc instance",
+			ledgerRange.FirstLedger.Sequence,
+			ledgerRange.LastLedger.Sequence,
+		)
+	}
+
+	return IsValidFormat(req.Format)
+}
+
+// LedgerInfo represents a single ledger in the response.
+type LedgerInfo struct {
+	Hash            string `json:"hash"`
+	Sequence        uint32 `json:"sequence"`
+	LedgerCloseTime int64  `json:"ledgerCloseTime,string"`
+
+	LedgerHeader     string          `json:"headerXdr"`
+	LedgerHeaderJSON json.RawMessage `json:"headerJson,omitempty"`
+
+	LedgerMetadata     string          `json:"metadataXdr"`
+	LedgerMetadataJSON json.RawMessage `json:"metadataJson,omitempty"`
+}
+
+// GetLedgersResponse encapsulates the response structure for getLedgers queries.
+type GetLedgersResponse struct {
+	Ledgers               []LedgerInfo `json:"ledgers"`
+	LatestLedger          uint32       `json:"latestLedger"`
+	LatestLedgerCloseTime int64        `json:"latestLedgerCloseTime"`
+	OldestLedger          uint32       `json:"oldestLedger"`
+	OldestLedgerCloseTime int64        `json:"oldestLedgerCloseTime"`
+	Cursor                string       `json:"cursor"`
+}
+
+type ledgersHandler struct {
+	ledgerReader db.LedgerReader
+	maxLimit     uint
+	defaultLimit uint
+}
+
+// NewGetLedgersHandler returns a jrpc2.Handler for the getLedgers method.
+func NewGetLedgersHandler(ledgerReader db.LedgerReader, maxLimit, defaultLimit uint) jrpc2.Handler {
+	return NewHandler((&ledgersHandler{
+		ledgerReader: ledgerReader,
+		maxLimit:     maxLimit,
+		defaultLimit: defaultLimit,
+	}).getLedgers)
+}
+
+// getLedgers fetch ledgers and relevant metadata from DB.
+func (h ledgersHandler) getLedgers(ctx context.Context, request GetLedgersRequest) (GetLedgersResponse, error) {
+	ledgerRange, err := h.ledgerReader.GetLedgerRange(ctx)
+	if err != nil {
+		return GetLedgersResponse{}, &jrpc2.Error{
+			Code:    jrpc2.InternalError,
+			Message: err.Error(),
+		}
+	}
+
+	if err := request.validate(h.maxLimit, ledgerRange); err != nil {
+		return GetLedgersResponse{}, &jrpc2.Error{
+			Code:    jrpc2.InvalidRequest,
+			Message: err.Error(),
+		}
+	}
+
+	start, limit, err := h.initializePagination(request, ledgerRange)
+	if err != nil {
+		return GetLedgersResponse{}, &jrpc2.Error{
+			Code:    jrpc2.InvalidParams,
+			Message: err.Error(),
+		}
+	}
+
+	ledgers, err := h.fetchLedgers(ctx, start, limit, request.Format)
+	if err != nil {
+		return GetLedgersResponse{}, err
+	}
```

Comment:
`GetLedgerRange` and `fetchLedgers` should be run within the same transaction in order to avoid races (e.g. AFAIU a new ledger could be ingested between `GetLedgerRange` and `fetchLedgers`). I don't know if we are doing the same in other endpoints, but we should. If not, please create a ticket for it.

Reply:
BTW, I am happy to be proven wrong if you think that the race I am mentioning cannot happen.

Reply:
getEvents addressed this concern without a transaction by reducing `limit` in the case that `start + limit` is greater than `ledgerRange.LastLedger.Sequence`

Reply:
Fair enough, then we should do that here too

Reply:
@2opremio Yeah nice catch. I think @tamirms suggestion would work here. Will add it and also add a test for it

Reply:
Actually I just discussed this with @tamirms and just validating the `start + limit` value is not enough. It would still lead to a race in the following scenario:
1. `start` = `oldestLedger` when the `GetLedgerRange` is called.
2. Before `fetchLedgers` is called, a new ledger is ingested and `start` is trimmed off.
3. `fetchLedgers` for the `start` ledger will fail since it is not present in the db.

So I think a common transaction approach will need to be implemented here. Looks like `getEvents` also suffers from the same problem 

cc @2opremio @Shaptic